web/src/search/stream.tsx

```diff
@@ -0,0 +1,159 @@
+// The protocol and implementation is still in the design phase. Feel free to
+// change anything and everything here. We are iteratively improving this
+// until it is no longer a proof of concept and instead works well.
+
+type SearchEvent = FileMatch[]
+
+interface FileMatch {
+    name: string
+    repository: string
+    branches?: [string]
+    version?: string
+    lineMatches: [LineMatch]
+}
+
+interface LineMatch {
+    line: string
+    lineNumber: number
+    offsetAndLengths: [[number]]
+}
+
+const toGQLLineMatch = (line: LineMatch): GQL.ILineMatch => ({
+    __typename: 'LineMatch',
+    limitHit: false,
+    lineNumber: line.lineNumber,
+    offsetAndLengths: line.offsetAndLengths,
+    preview: line.line,
+})
+
+function toGQLFileMatch(fm: FileMatch): GQL.IFileMatch {
+    let revision = ''
+    if (fm.branches) {
+        const branch = fm.branches[0]
+        if (branch !== '') {
+            revision = '@' + branch
+        }
+    } else if (fm.version) {
+        revision = '@' + fm.version
+    }
+
+    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+    const file: GQL.IGitBlob = {
+        path: fm.name,
+        // /github.com/gorilla/mux@v1.7.2/-/blob/mux_test.go
+        // TODO return in response?
+        url: '/' + fm.repository + revision + '/-/blob/' + fm.name,
+        commit: {
+            oid: fm.version || '',
+        },
+    } as GQL.IGitBlob
+    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+    const repository: GQL.IRepository = {
+        name: fm.repository,
+    } as GQL.IRepository
+    return {
+        __typename: 'FileMatch',
+        file,
+        repository,
+        revSpec: null,
+        resource: fm.name,
+        symbols: [],
+        lineMatches: fm.lineMatches.map(toGQLLineMatch),
+        limitHit: false,
+    }
+}
+
+// TODO fill in the fields we actually care about
+const toGQLSearchResults = (results: GQL.SearchResult[]): GQL.ISearchResults => ({
+    __typename: 'SearchResults',
+    matchCount: 0,
+    resultCount: 0,
+    approximateResultCount: '',
+    limitHit: false,
+    sparkline: [],
+    repositories: [],
+    repositoriesCount: 0,
+    repositoriesSearched: [],
+    indexedRepositoriesSearched: [],
+    cloning: [],
+    missing: [],
+    timedout: [],
+    indexUnavailable: false,
+    alert: null,
+    elapsedMilliseconds: 0,
+    dynamicFilters: [],
+    results,
+    pageInfo: { __typename: 'PageInfo', endCursor: null, hasNextPage: false },
+})
+
+/**
+ * Converts a stream of SearchEvents into an aggregated GQL.ISearchResult
+ */
+export function switchToGQLISearchResults(): OperatorFunction<SearchEvent, GQL.ISearchResults> {
+    let allFileMatches: GQL.IFileMatch[] = []
+
+    return map<SearchEvent, GQL.ISearchResults>(fileMatches => {
+        allFileMatches = allFileMatches.concat(fileMatches.map(toGQLFileMatch))
+
+        return toGQLSearchResults(allFileMatches)
+    })
+}
```

Comment:
Nicer declarative alternative

```suggestion
export const switchToGQLISearchResults: OperatorFunction<SearchEvent, GQL.ISearchResults> = searchEvents =>
    searchEvents.pipe(
        map(fileMatches => fileMatches.map(toGQLFileMatch)),
        scan((allFileMatches, newFileMatches) => allFileMatches.concat(newFileMatches), []),
        map(toGQLSearchResults)
    )
```